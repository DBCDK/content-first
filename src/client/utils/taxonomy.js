import {getTaxonomy} from '../../shared/taxonomy.requester';

const {difference} = require('lodash');
const taxonomy = getTaxonomy();

const getLeaves = (t, parentStack = []) => {
  t = t ? t : getTaxonomy();
  if (Array.isArray(t)) {
    return t.map(leaf => {
      return Object.assign({}, leaf, {parents: [...parentStack]});
    });
  }
  let result = [];
  Object.entries(t).forEach(([key, value]) => {
    parentStack.push(key);
    const tmp = getLeaves(value, parentStack);
    parentStack.pop();
    result = result.concat(tmp);
  });
  return result;
};

const getLeavesMap = t => {
  t = t ? t : getTaxonomy();
  const res = {};
  getLeaves(t).forEach(leaf => {
    res[leaf.id] = leaf;
  });
  return res;
};

const getFromTitleMap = t => {
  t = t ? t : getTaxonomy();
  const res = {};
  getLeaves(t).forEach(leaf => {
    if (!leaf.parents.includes('Hovedperson(er)')) {
      res[leaf.title.toLowerCase()] = leaf;
    }
  });
  return res;
};

const isRange = tagId => Array.isArray(tagId) && tagId.length === 2;
const isFullRange = (tagId, fullRange) =>
  fullRange[0] === tagId[0] && fullRange[fullRange.length - 1] === tagId[1];

const isSameRange = (tagId, fullRange) =>
  isRange(tagId) && difference(tagId, fullRange).length === 0;
const getFullRange = (tagId, filterCards, filtersMapAll) => {
  const filter = filtersMapAll[tagId] || filtersMapAll[tagId[0]];
  if (!filter) {
    return false;
  }
  const parent = filtersMapAll[filter.id].parents[0];
  const range = (filterCards[parent] && filterCards[parent].range) || false;
  return range;
};
const findRangeLocation = (tagId, tags = [], fullRange) => {
  for (let i = 0; i < tags.length; i++) {
    if (isSameRange(tags[i], fullRange)) {
      return i;
    }
  }
  return -1;
};
const getDistances = (tagId, fullRange) => {
  const index = fullRange.indexOf(tagId);
  return {begin: index, end: fullRange.length - index};
};
const getQueryType = (selectedTagIdss, selectedCreators, selectedTitles) => {
  if (selectedTagIdss.length > 0) {
    return 'tags';
  }
  if (selectedCreators.length > 0 || selectedTitles.length > 0) {
    return 'titlecreator';
  }
  return false;
};
const getSelectedRange = (tagId, selectedRange, fullRange) => {
  if (isRange(tagId)) {
    return tagId;
  }
  if (!selectedRange) {
    return [tagId, tagId];
  }
  if (fullRange.indexOf(tagId) <= fullRange.indexOf(selectedRange[0])) {
    return [tagId, selectedRange[1]];
  } else if (fullRange.indexOf(tagId) >= fullRange.indexOf(selectedRange[0])) {
    return [selectedRange[0], tagId];
  }
  const distances = getDistances(tagId, fullRange);
  if (distances.begin < distances.end) {
    return [tagId, selectedRange[1]];
  }
  return [selectedRange[0], tagId];
};
const tagsToUrlParams = tags =>
  tags.map(id => (isRange(id) ? id.join(':') : id)).join(',');

export {
  taxonomy,
  getLeaves,
  getLeavesMap,
  getFromTitleMap,
  isRange,
  isFullRange,
  isSameRange,
  getFullRange,
  findRangeLocation,
  getDistances,
  getQueryType,
  getSelectedRange,
  tagsToUrlParams
};
