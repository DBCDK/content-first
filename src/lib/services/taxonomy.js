const _ = require('lodash');
const request = require('superagent');

const exportTaxonomyURL =
  'https://artifactory.dbc.dk/artifactory/fe-generic/metakompasset/json-files.tar.gz!/exportTaxonomy.json';

class Taxonomy {
  constructor(config, logger) {
    this.config = config;
    this.logger = logger;
    this.clear();
  }

  getName() {
    return 'taxonomy';
  }

  clear() {
    this.setOk();
    this.errorLog = [];
  }

  setOk() {
    this.ok = true;
    this.currentError = null;
  }

  isOk() {
    return this.ok;
  }

  getCurrentError() {
    return this.currentError;
  }

  getErrorLog() {
    return this.errorLog;
  }

  async getTaxonomy() {
    try {
      const result = await request.get(exportTaxonomyURL);
      return result.body;
    } catch (e) {
      const msg = _.get(e, 'response.body.value') || 'Internal server error';
      this.logger.log.error('get suggestions - error', {
        source: 'taxonomy',
        errorMessage: msg,
        stack: e.stack
      });
      throw new Error(msg);
    }
  }
}

module.exports = Taxonomy;
