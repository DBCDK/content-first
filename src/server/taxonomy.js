/**
 * Utillties for backend service manipulation of taxonomy data.
 */

'use strict';
const _ = require('lodash');
const {validatingInput} = require('__/json');
const path = require('path');
const schema = path.join(__dirname, 'schemas/taxonomy-in.json');

class IdChecker {
  constructor() {
    this.problems = [];
    this.ids = [];
  }
  getProblems() {
    return this.problems;
  }
  checkId(x) {
    const id = parseInt(x, 10);
    if (_.isNaN(id)) {
      this.problems.push(`Cannot convert ${x} to an integer`);
      return;
    }
    if (_.includes(this.ids, id)) {
      this.problems.push(`Id ${id} occurs more than once`);
      return;
    }
    this.ids.push(id);
  }
}

function parsingTaxonomyInjection(obj) {
  return new Promise((resolve, reject) => {
    return validatingInput(obj, schema)
      .then(document => {
        const idChecker = new IdChecker();
        document.forEach(top => {
          idChecker.checkId(top.id);
          top.items.forEach(middle => {
            idChecker.checkId(middle.id);
            middle.items.forEach(bottom => {
              idChecker.checkId(bottom.id);
            });
          });
        });
        const problems = idChecker.getProblems();
        if (problems.length !== 0) {
          return reject(problems);
        }
        let taxonomy = [];
        document.forEach(top => {
          let middleTier = [];
          top.items.forEach(middle => {
            let bottomTier = [];
            middle.items.forEach(bottom => {
              bottomTier.push({
                id: parseInt(bottom.id, 10),
                title: bottom.title
              });
            });
            middleTier.push({
              id: parseInt(middle.id, 10),
              title: middle.title,
              items: bottomTier
            });
          });
          taxonomy.push({
            id: parseInt(top.id, 10),
            title: top.title,
            items: middleTier
          });
        });
        resolve(taxonomy);
      })
      .catch(reject);
  });
}

exports.parsingTaxonomyInjection = parsingTaxonomyInjection;
