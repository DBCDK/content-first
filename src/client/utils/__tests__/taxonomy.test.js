import {getLeaves, getLeavesMap} from '../taxonomy';

const taxonomy = {
  a: {
    aa: [{id: 1, title: 'leaf1', score: 0}, {id: 2, title: 'leaf2', score: 0}, {id: 3, title: 'leaf3', score: 0}]
  },
  b: [{id: 4, title: 'leaf4', score: 0}, {id: 5, title: 'leaf5', score: 0}],
  c: {cc: {ccc: [{id: 6, title: 'leaf6', score: 0}, {id: 7, title: 'leaf7', score: 0}]}}
};

describe('getLeaves', () => {
  test('Get leaves of the tree', () => {
    expect(getLeaves(taxonomy)).toMatchSnapshot();
  });

  test('Get leaves of the tree as a map', () => {
    expect(getLeavesMap(taxonomy)).toMatchSnapshot();
  });
});
