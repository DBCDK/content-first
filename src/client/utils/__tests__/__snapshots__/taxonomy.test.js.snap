// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getLeaves Get leaves of the tree 1`] = `
Array [
  Object {
    "id": 1,
    "parents": Array [
      "a",
      "aa",
    ],
    "score": 0,
    "title": "leaf1",
  },
  Object {
    "id": 2,
    "parents": Array [
      "a",
      "aa",
    ],
    "score": 0,
    "title": "leaf2",
  },
  Object {
    "id": 3,
    "parents": Array [
      "a",
      "aa",
    ],
    "score": 0,
    "title": "leaf3",
  },
  Object {
    "id": 4,
    "parents": Array [
      "b",
    ],
    "score": 0,
    "title": "leaf4",
  },
  Object {
    "id": 5,
    "parents": Array [
      "b",
    ],
    "score": 0,
    "title": "leaf5",
  },
  Object {
    "id": 6,
    "parents": Array [
      "c",
      "cc",
      "ccc",
    ],
    "score": 0,
    "title": "leaf6",
  },
  Object {
    "id": 7,
    "parents": Array [
      "c",
      "cc",
      "ccc",
    ],
    "score": 0,
    "title": "leaf7",
  },
  Object {
    "id": 8,
    "parents": Array [
      "Hovedperson(er)",
    ],
    "score": 0,
    "title": "leaf8",
  },
]
`;

exports[`getLeaves Get leaves of the tree as a map 1`] = `
Object {
  "1": Object {
    "id": 1,
    "parents": Array [
      "a",
      "aa",
    ],
    "score": 0,
    "title": "leaf1",
  },
  "2": Object {
    "id": 2,
    "parents": Array [
      "a",
      "aa",
    ],
    "score": 0,
    "title": "leaf2",
  },
  "3": Object {
    "id": 3,
    "parents": Array [
      "a",
      "aa",
    ],
    "score": 0,
    "title": "leaf3",
  },
  "4": Object {
    "id": 4,
    "parents": Array [
      "b",
    ],
    "score": 0,
    "title": "leaf4",
  },
  "5": Object {
    "id": 5,
    "parents": Array [
      "b",
    ],
    "score": 0,
    "title": "leaf5",
  },
  "6": Object {
    "id": 6,
    "parents": Array [
      "c",
      "cc",
      "ccc",
    ],
    "score": 0,
    "title": "leaf6",
  },
  "7": Object {
    "id": 7,
    "parents": Array [
      "c",
      "cc",
      "ccc",
    ],
    "score": 0,
    "title": "leaf7",
  },
  "8": Object {
    "id": 8,
    "parents": Array [
      "Hovedperson(er)",
    ],
    "score": 0,
    "title": "leaf8",
  },
}
`;

exports[`getLeaves Get leaves the title map 1`] = `
Object {
  "leaf1": Object {
    "id": 1,
    "parents": Array [
      "a",
      "aa",
    ],
    "score": 0,
    "title": "leaf1",
  },
  "leaf2": Object {
    "id": 2,
    "parents": Array [
      "a",
      "aa",
    ],
    "score": 0,
    "title": "leaf2",
  },
  "leaf3": Object {
    "id": 3,
    "parents": Array [
      "a",
      "aa",
    ],
    "score": 0,
    "title": "leaf3",
  },
  "leaf4": Object {
    "id": 4,
    "parents": Array [
      "b",
    ],
    "score": 0,
    "title": "leaf4",
  },
  "leaf5": Object {
    "id": 5,
    "parents": Array [
      "b",
    ],
    "score": 0,
    "title": "leaf5",
  },
  "leaf6": Object {
    "id": 6,
    "parents": Array [
      "c",
      "cc",
      "ccc",
    ],
    "score": 0,
    "title": "leaf6",
  },
  "leaf7": Object {
    "id": 7,
    "parents": Array [
      "c",
      "cc",
      "ccc",
    ],
    "score": 0,
    "title": "leaf7",
  },
}
`;
